src/optimizer/optimizer_task.cpp

```diff
@@ -0,0 +1,467 @@
+          cur_total_cost_ += context_->GetOptimizerContext()->GetCostModel()->CalculateCost(
+              context_->GetOptimizerContext()->GetTxn(), &context_->GetOptimizerContext()->GetMemo(),
+              memo_enforced_expr);
+
+          // Update hash tables for group and group expression
+          memo_enforced_expr->SetLocalHashTable(extended_prop_set, {pre_output_prop_set}, cur_total_cost_);
+          cur_group->SetExpressionCost(memo_enforced_expr, cur_total_cost_, output_prop->Copy());
+        }
+      }
+
+      // Can meet the requirement
+      if (meet_requirement && cur_total_cost_ <= context_->GetCostUpperBound()) {
+        // If the cost is smaller than the winner, update the context upper bound
+        context_->SetCostUpperBound(context_->GetCostUpperBound() - cur_total_cost_);
+        if (memo_enforced_expr != nullptr) {  // Enforcement takes place
+          cur_group->SetExpressionCost(memo_enforced_expr, cur_total_cost_, context_->GetRequiredProperties()->Copy());
+        } else if (output_prop->Properties().size() != context_->GetRequiredProperties()->Properties().size()) {
+          // The original output property is a super set of the requirement
+          cur_group->SetExpressionCost(group_expr_, cur_total_cost_, context_->GetRequiredProperties()->Copy());
+        }
+      }
+    }
+
+    // Reset child idx and total cost
+    prev_child_idx_ = -1;
+    cur_child_idx_ = 0;
+    cur_total_cost_ = 0;
+  }
+}
+
+void TopDownRewrite::Execute() {
+  std::vector<RuleWithPromise> valid_rules;
+
+  auto cur_group = GetMemo().GetGroupByID(group_id_);
+  auto cur_group_expr = cur_group->GetLogicalExpression();
+
+  // Construct valid transformation rules from rule set
+  std::vector<Rule *> set = GetRuleSet().GetRewriteRulesByName(rule_set_name_);
+  ConstructValidRules(cur_group_expr, context_, set, &valid_rules);
+
+  // Sort so that we apply rewrite rules with higher promise first
+  std::sort(valid_rules.begin(), valid_rules.end(), std::greater<>());
+
+  for (auto &r : valid_rules) {
+    Rule *rule = r.GetRule();
+    GroupExprBindingIterator iterator(GetMemo(), cur_group_expr, rule->GetMatchPattern());
+    if (iterator.HasNext()) {
+      auto before = iterator.Next();
+      TERRIER_ASSERT(!iterator.HasNext(), "there should only be 1 binding");
+      std::vector<std::unique_ptr<OperatorExpression>> after;
+      rule->Transform(common::ManagedPointer(before.get()), &after, context_);
+
+      // Rewrite rule should provide at most 1 expression
+      TERRIER_ASSERT(after.size() <= 1, "rule provided too many transformations");
+      if (!after.empty()) {
+        auto &new_expr = after[0];
+        context_->GetOptimizerContext()->ReplaceRewriteExpression(common::ManagedPointer(new_expr.get()), group_id_);
+        PushTask(new TopDownRewrite(group_id_, context_, rule_set_name_));
+        return;
+      }
+    }
+
+    cur_group_expr->SetRuleExplored(rule);
```

Comment:
It is possible since `ConstructValidRules` does not do a full pattern binding check. In another scenario, `Rule::Transform` might not have created a transformed `OperatorExpression`.